Fit list Impress feature - dynamically adapts font size for outline text to fit the shape area

From: Thorsten Behrens <thb@openoffice.org>


---

 offapi/com/sun/star/drawing/TextFitToSizeType.idl  |    7 +
 qadevOOo/runner/util/ValueChanger.java             |    2 
 sd/sdi/_drvwsh.sdi                                 |   12 --
 svx/inc/svx/sdr/attribute/sdrtextattribute.hxx     |    6 +
 svx/inc/svx/sdr/primitive2d/sdrtextprimitive2d.hxx |   40 ++++++
 .../svx/sdr/primitive2d/svx_primitivetypes2d.hxx   |    1 
 svx/inc/svx/sdtfsitm.hxx                           |   26 ++--
 svx/inc/svx/svdotext.hxx                           |   16 ++
 svx/source/dialog/dbregisterednamesconfig.cxx      |    4 -
 svx/source/dialog/textattr.cxx                     |    1 
 svx/source/editeng/editobj2.hxx                    |   10 +
 svx/source/editeng/impedit3.cxx                    |    7 +
 svx/source/editeng/impedit4.cxx                    |    6 +
 svx/source/outliner/outliner.cxx                   |   11 +-
 svx/source/sdr/attribute/sdrtextattribute.cxx      |    3 
 svx/source/sdr/primitive2d/sdrattributecreator.cxx |    6 -
 .../sdr/primitive2d/sdrdecompositiontools.cxx      |    5 +
 svx/source/sdr/primitive2d/sdrtextprimitive2d.cxx  |   54 ++++++++
 svx/source/svdraw/svdedxv.cxx                      |   10 +
 svx/source/svdraw/svdfppt.cxx                      |   12 ++
 svx/source/svdraw/svdotext.cxx                     |  120 ++++++++++++++---
 svx/source/svdraw/svdotextdecomposition.cxx        |  140 +++++++++++++++++++-
 svx/source/svdraw/svdotxat.cxx                     |    3 
 svx/source/svdraw/svdotxed.cxx                     |   17 ++
 svx/source/svdraw/svdotxtr.cxx                     |    9 -
 svx/source/svdraw/svdtxhdl.cxx                     |    3 
 svx/source/svdraw/svdview.cxx                      |    3 
 xmloff/source/draw/sdpropls.cxx                    |    4 -
 28 files changed, 440 insertions(+), 98 deletions(-)


diff --git offapi/com/sun/star/drawing/TextFitToSizeType.idl offapi/com/sun/star/drawing/TextFitToSizeType.idl
index d17a737..d33a674 100644
--- offapi/com/sun/star/drawing/TextFitToSizeType.idl
+++ offapi/com/sun/star/drawing/TextFitToSizeType.idl
@@ -63,9 +63,10 @@ published enum TextFitToSizeType
  
 	//------------------------------------------------------------------------- 
 
-	/** if the shape is scaled, the font attributes are scaled and hard set
-		on the text */
-	RESIZEATTR 
+	/** if the shape is scaled, the font is scaled isotrophically to
+		fit the avaiable space. Auto line-breaks will keep working
+   */
+	AUTOFIT 
  
 }; 
  
diff --git qadevOOo/runner/util/ValueChanger.java qadevOOo/runner/util/ValueChanger.java
index 32a295a..faeae99 100644
--- qadevOOo/runner/util/ValueChanger.java
+++ qadevOOo/runner/util/ValueChanger.java
@@ -389,7 +389,7 @@ public class ValueChanger {
         com.sun.star.drawing.TextFitToSizeType TF1 = com.sun.star.drawing.TextFitToSizeType.ALLLINES;
         com.sun.star.drawing.TextFitToSizeType TF2 = com.sun.star.drawing.TextFitToSizeType.NONE;
         com.sun.star.drawing.TextFitToSizeType TF3 = com.sun.star.drawing.TextFitToSizeType.PROPORTIONAL;
-        com.sun.star.drawing.TextFitToSizeType TF4 = com.sun.star.drawing.TextFitToSizeType.RESIZEATTR;
+        com.sun.star.drawing.TextFitToSizeType TF4 = com.sun.star.drawing.TextFitToSizeType.AUTOFIT;
         if (oldValue.equals(TF1)) newValue = TF2;
         if (oldValue.equals(TF2)) newValue = TF3;
         if (oldValue.equals(TF3)) newValue = TF4;
diff --git sd/sdi/_drvwsh.sdi sd/sdi/_drvwsh.sdi
index a3dfd27..c895d2e 100644
--- sd/sdi/_drvwsh.sdi
+++ sd/sdi/_drvwsh.sdi
@@ -28,18 +28,6 @@
  *
  ************************************************************************/
 
-enum SdrFitToSizeType
-{
-    SDRTEXTFIT_NONE ,
-    SDRTEXTFIT_PROPORTIONAL ,
-    SDRTEXTFIT_ALLLINES ,
-    SDRTEXTFIT_RESIZEATTR
-}
-item UINT32 SvxObjectItem ;
-item BOOL SdrShadowItem ;
-item SdrFitToSizeType SdrTextFitToSizeTypeItem ;
-
-
 interface DrawView
 {
 	SID_JUMPTOMARK // ole : no, status : ?
diff --git svx/inc/svx/sdr/attribute/sdrtextattribute.hxx svx/inc/svx/sdr/attribute/sdrtextattribute.hxx
index 713ec55..d324e8e 100644
--- svx/inc/svx/sdr/attribute/sdrtextattribute.hxx
+++ svx/inc/svx/sdr/attribute/sdrtextattribute.hxx
@@ -64,6 +64,7 @@ namespace drawinglayer
 			// bitfield
 			unsigned								mbContour : 1;
 			unsigned								mbFitToSize : 1;
+			unsigned								mbAutoFit : 1;
 			unsigned								mbHideContour : 1;
 			unsigned								mbBlink : 1;
 			unsigned								mbScroll : 1;
@@ -77,8 +78,8 @@ namespace drawinglayer
 				sal_Int32 aTextRightDistance, 
 				sal_Int32 aTextLowerDistance, 
 				bool bContour, bool bFitToSize, 
-				bool bHideContour, bool bBlink, 
-				bool bScroll);
+                bool bAutoFit, bool bHideContour, 
+                bool bBlink, bool bScroll);
 			bool operator==(const SdrTextAttribute& rCandidate) const;
 
 			// data access
@@ -86,6 +87,7 @@ namespace drawinglayer
 			bool isContour() const { return mbContour; }
 			bool isFontwork() const { return (XFT_NONE != meFormTextStyle); }
 			bool isFitToSize() const { return mbFitToSize; }
+			bool isAutoFit() const { return mbAutoFit; }
 			bool isHideContour() const { return mbHideContour; }
 			bool isBlink() const { return mbBlink; }
 			bool isScroll() const { return mbScroll; }
diff --git svx/inc/svx/sdr/primitive2d/sdrtextprimitive2d.hxx svx/inc/svx/sdr/primitive2d/sdrtextprimitive2d.hxx
index 78b9210..8432d40 100644
--- svx/inc/svx/sdr/primitive2d/sdrtextprimitive2d.hxx
+++ svx/inc/svx/sdr/primitive2d/sdrtextprimitive2d.hxx
@@ -251,6 +251,46 @@ namespace drawinglayer
 
 //////////////////////////////////////////////////////////////////////////////
 
+namespace drawinglayer
+{
+	namespace primitive2d
+	{
+		class SdrAutoFitTextPrimitive2D : public SdrTextPrimitive2D
+		{
+		private:
+			::basegfx::B2DHomMatrix					maTextRangeTransform;	// text range transformation from unit range ([0.0 .. 1.0]) to text range
+
+			// bitfield
+            unsigned                                mbWordWrap : 1;         // for CustomShapes text layout
+
+		protected:
+			// local decomposition.
+			virtual Primitive2DSequence createLocalDecomposition(const geometry::ViewInformation2D& aViewInformation) const;
+
+		public:
+			SdrAutoFitTextPrimitive2D(
+				const SdrText& rSdrText,
+				const ::basegfx::B2DHomMatrix& rTextRangeTransform,
+                bool bWordWrap);
+
+			// get data
+			const basegfx::B2DHomMatrix& getTextRangeTransform() const { return maTextRangeTransform; }
+			bool getWordWrap() const { return mbWordWrap; }
+
+			// compare operator
+			virtual bool operator==(const BasePrimitive2D& rPrimitive) const;
+
+			// transformed clone operator
+			virtual SdrTextPrimitive2D* createTransformedClone(const ::basegfx::B2DHomMatrix& rTransform) const;
+
+			// provide unique ID
+			DeclPrimitrive2DIDBlock()
+		};
+	} // end of namespace primitive2d
+} // end of namespace drawinglayer
+
+//////////////////////////////////////////////////////////////////////////////
+
 #endif //INCLUDED_SDR_PRIMITIVE2D_SDRTEXTPRIMITIVE2D_HXX
 
 // eof
diff --git svx/inc/svx/sdr/primitive2d/svx_primitivetypes2d.hxx svx/inc/svx/sdr/primitive2d/svx_primitivetypes2d.hxx
index b0f0e62..5ce667c 100644
--- svx/inc/svx/sdr/primitive2d/svx_primitivetypes2d.hxx
+++ svx/inc/svx/sdr/primitive2d/svx_primitivetypes2d.hxx
@@ -52,6 +52,7 @@
 #define PRIMITIVE2D_ID_SDRSTRETCHTEXTPRIMITIVE2D		(PRIMITIVE2D_ID_RANGE_SVX| 13)
 #define PRIMITIVE2D_ID_SDRCELLPRIMITIVE2D				(PRIMITIVE2D_ID_RANGE_SVX| 14)
 #define PRIMITIVE2D_ID_SDRBORDERLINEPRIMITIVE2D			(PRIMITIVE2D_ID_RANGE_SVX| 15)
+#define PRIMITIVE2D_ID_SDRAUTOFITTEXTPRIMITIVE2D		(PRIMITIVE2D_ID_RANGE_SVX| 16)
 
 //////////////////////////////////////////////////////////////////////////////
 
diff --git svx/inc/svx/sdtfsitm.hxx svx/inc/svx/sdtfsitm.hxx
index 97fb158..474cf75 100644
--- svx/inc/svx/sdtfsitm.hxx
+++ svx/inc/svx/sdtfsitm.hxx
@@ -34,19 +34,21 @@
 #include <svx/svddef.hxx>
 #include "svx/svxdllapi.h"
 
-enum SdrFitToSizeType {SDRTEXTFIT_NONE,         // - kein FitToSize
-					   SDRTEXTFIT_PROPORTIONAL, // - Alle Buchstaben proportional umgroessern
-					   SDRTEXTFIT_ALLLINES,     // - Zus. jede Zeile separat in der Breite stretchen
-					   SDRTEXTFIT_RESIZEATTR};  // - Bei Rahmenumgroesserung (ausser Autogrow) wird
-												//   die Schriftgroesse umattributiert (hart)
+enum SdrFitToSizeType {
+    SDRTEXTFIT_NONE,         // - no fit-to-size
+    SDRTEXTFIT_PROPORTIONAL, // - resize all glyhs proportionally
+                             //   (might scale anisotrophically)
+    SDRTEXTFIT_ALLLINES,     // - like SDRTEXTFIT_PROPORTIONAL, but
+                             //   scales each line separately
+    SDRTEXTFIT_AUTOFIT};     // - mimics PPT's automatic adaption of
+                             //   font size to text rect - comparable
+                             //   to SDRTEXTFIT_PROPORTIONAL, but
+                             //   scales isotrophically
 
-// Bei SDRTEXTFIT_PROPORTIONAL und SDRTEXTFIT_ALLLINES gibt es kein AutoGrow und
-// keine automatischen Umbrueche.
-// Ist SDRTEXTFIT_RESIZEATTR gesetzt, so wird beim umgroessern des Textrahmens
-// (ausser bei AutoGrow) die Schrift durch harte Attributierung ebenfalls
-// umgegroessert.
-// Bei AutoGrowingWidth gibt es ebenfalls keine automatischen Umbrueche (erst bei
-// TextMaxFrameWidth).
+// No AutoGrow and no automatic line breaks for
+// SDRTEXTFIT_PROPORTIONAL and SDRTEXTFIT_ALLLINES.
+// No automatic line breaks for AutoGrowingWidth as well (only if
+// TextMaxFrameWidth is reached).
 
 //--------------------------------
 // class SdrTextFitToSizeTypeItem
diff --git svx/inc/svx/svdotext.hxx svx/inc/svx/svdotext.hxx
index a1ce415..91a7fd9 100644
--- svx/inc/svx/svdotext.hxx
+++ svx/inc/svx/svdotext.hxx
@@ -62,6 +62,7 @@ namespace drawinglayer { namespace primitive2d {
 	class SdrContourTextPrimitive2D;
 	class SdrPathTextPrimitive2D;
 	class SdrBlockTextPrimitive2D;
+	class SdrAutoFitTextPrimitive2D;
 	class SdrStretchTextPrimitive2D;
 }}
 
@@ -259,6 +260,9 @@ protected:
 	// Flag for allowing text animation. Default is sal_true.
 	BOOL						mbTextAnimationAllowed : 1;
 
+    // flag for preventing recursive onEditOutlinerStatusEvent calls
+	BOOL					    mbInDownScale : 1;
+
 	SVX_DLLPRIVATE SdrOutliner& ImpGetDrawOutliner() const;
 
 private:
@@ -272,6 +276,8 @@ private:
                                        Rectangle& 		rAnchorRect, 
                                        Rectangle& 		rPaintRect, 
                                        Fraction& 		aFitXKorreg ) const;
+    void ImpAutoFitText( SdrOutliner& rOutliner ) const;
+    static void ImpAutoFitText( SdrOutliner& rOutliner, const Size& rShapeSize, bool bIsVerticalWriting );
 	SVX_DLLPRIVATE SdrObject* ImpConvertObj(FASTBOOL bToPoly) const;
 	SVX_DLLPRIVATE void ImpLinkAnmeldung();
 	SVX_DLLPRIVATE void ImpLinkAbmeldung();
@@ -283,7 +289,7 @@ protected:
 	SdrObject* ImpConvertMakeObj(const basegfx::B2DPolyPolygon& rPolyPolygon, sal_Bool bClosed, sal_Bool bBezier, sal_Bool bNoSetAttr = sal_False) const;
 	SdrObject* ImpConvertAddText(SdrObject* pObj, FASTBOOL bBezier) const;
 	void ImpSetTextStyleSheetListeners();
-	void ImpSetCharStretching(SdrOutliner& rOutliner, const Rectangle& rTextRect, const Rectangle& rAnchorRect, Fraction& rFitXKorreg) const;
+    void ImpSetCharStretching(SdrOutliner& rOutliner, const Size& rTextSize, const Size& rShapeSize, Fraction& rFitXKorreg) const;
 	void ImpJustifyRect(Rectangle& rRect) const;
 	void ImpCheckShear();
 	Rectangle ImpDragCalcRect(const SdrDragStat& rDrag) const;
@@ -345,6 +351,10 @@ public:
 	void NbcResizeTextAttributes(const Fraction& xFact, const Fraction& yFact);
 	FASTBOOL IsTextFrame() const { return bTextFrame; }
 	FASTBOOL IsOutlText() const { return bTextFrame && (eTextKind==OBJ_OUTLINETEXT || eTextKind==OBJ_TITLETEXT); }
+    /// returns true if the PPT autofit of text into shape bounds is enabled. implies IsFitToSize()==false!
+	FASTBOOL IsAutoFit() const;
+    /// returns true if the old feature for fitting shape content should into shape is enabled. implies IsAutoFit()==false!
+	FASTBOOL IsFitToSize() const;
 	SdrObjKind GetTextKind() const { return eTextKind; }
 
 	virtual bool HasText() const;
@@ -587,6 +597,10 @@ public:
 		drawinglayer::primitive2d::Primitive2DSequence& rTarget, 
 		const drawinglayer::primitive2d::SdrBlockTextPrimitive2D& rSdrBlockTextPrimitive,
 		const drawinglayer::geometry::ViewInformation2D& aViewInformation) const;
+	bool impDecomposeAutoFitTextPrimitive(
+		drawinglayer::primitive2d::Primitive2DSequence& rTarget, 
+		const drawinglayer::primitive2d::SdrAutoFitTextPrimitive2D& rSdrAutofitTextPrimitive,
+		const drawinglayer::geometry::ViewInformation2D& aViewInformation) const;
 	bool impDecomposeStretchTextPrimitive(
 		drawinglayer::primitive2d::Primitive2DSequence& rTarget, 
 		const drawinglayer::primitive2d::SdrStretchTextPrimitive2D& rSdrStretchTextPrimitive,
diff --git svx/source/dialog/dbregisterednamesconfig.cxx svx/source/dialog/dbregisterednamesconfig.cxx
index 0723af6..567baef 100644
--- svx/source/dialog/dbregisterednamesconfig.cxx
+++ svx/source/dialog/dbregisterednamesconfig.cxx
@@ -163,7 +163,7 @@ namespace svx
                     }
                     catch( const Exception& )
                     {
-                    	DBG_UNHANDLED_EXCEPTION();
+                    	//DBG_UNHANDLED_EXCEPTION();
                     }
                 }
 			}
@@ -184,7 +184,7 @@ namespace svx
                     }
                     catch( const Exception& )
                     {
-                    	DBG_UNHANDLED_EXCEPTION();
+                    	//DBG_UNHANDLED_EXCEPTION();
                     }
                 }
 			}
diff --git svx/source/dialog/textattr.cxx svx/source/dialog/textattr.cxx
index 563a367..aa64e10 100644
--- svx/source/dialog/textattr.cxx
+++ svx/source/dialog/textattr.cxx
@@ -460,7 +460,6 @@ BOOL SvxTextAttrPage::FillItemSet( SfxItemSet& rAttrs)
             default: ; //prevent warning
 				DBG_ERROR( "svx::SvxTextAttrPage::FillItemSet(), unhandled state!" );
 			case STATE_NOCHECK: eFTS = SDRTEXTFIT_NONE; break;
-			//case STATE_CHECK: eFTS = SDRTEXTFIT_RESIZEATTR; break;
 			case STATE_CHECK: eFTS = SDRTEXTFIT_PROPORTIONAL; break;
 		}
 		rAttrs.Put( SdrTextFitToSizeTypeItem( eFTS ) );
diff --git svx/source/editeng/editobj2.hxx svx/source/editeng/editobj2.hxx
index 63ec23a..a44ad12 100644
--- svx/source/editeng/editobj2.hxx
+++ svx/source/editeng/editobj2.hxx
@@ -115,12 +115,16 @@ class XParaPortionList : public  XBaseParaPortionList
 	ULONG 		nRefDevPtr;
 	OutDevType	eRefDevType;
 	MapMode		aRefMapMode;
+	sal_uInt16	nStretchX;
+	sal_uInt16	nStretchY;
 	ULONG		nPaperWidth;
 
 
 public:
-			XParaPortionList( OutputDevice* pRefDev, ULONG nPW ) :
-				aRefMapMode( pRefDev->GetMapMode() )
+    XParaPortionList( OutputDevice* pRefDev, ULONG nPW, sal_uInt16 _nStretchX, sal_uInt16 _nStretchY ) :
+        aRefMapMode( pRefDev->GetMapMode() ),
+        nStretchX(_nStretchX),
+        nStretchY(_nStretchY)
 				{
 					nRefDevPtr = (ULONG)pRefDev; nPaperWidth = nPW;
 					eRefDevType = pRefDev->GetOutDevType();
@@ -130,6 +134,8 @@ public:
 	ULONG			GetPaperWidth() const 		{ return nPaperWidth; }
 	OutDevType		GetRefDevType() const 		{ return eRefDevType; }
 	const MapMode&	GetRefMapMode() const		{ return aRefMapMode; }
+	sal_uInt16	GetStretchX() const         { return nStretchX; }
+	sal_uInt16	GetStretchY() const         { return nStretchY; }
 };
 
 /* cl removed because not needed anymore since binfilter
diff --git svx/source/editeng/impedit3.cxx svx/source/editeng/impedit3.cxx
index fb8eac9..9391ed5 100644
--- svx/source/editeng/impedit3.cxx
+++ svx/source/editeng/impedit3.cxx
@@ -4030,20 +4030,25 @@ void ImpEditEngine::SetFlatMode( sal_Bool bFlat )
 
 void ImpEditEngine::SetCharStretching( sal_uInt16 nX, sal_uInt16 nY )
 {
+    bool bChanged(false);
 	if ( !IsVertical() )
 	{
+        bChanged = nStretchX!=nX || nStretchY!=nY;
 		nStretchX = nX;
 		nStretchY = nY;
 	}
 	else
 	{
+        bChanged = nStretchX!=nY || nStretchY!=nX;
 		nStretchX = nY;
 		nStretchY = nX;
 	}
 
-	if ( aStatus.DoStretch() )
+	if (bChanged && aStatus.DoStretch())
 	{
 		FormatFullDoc();
+        // (potentially) need everything redrawn
+        aInvalidRec=Rectangle(0,0,1000000,1000000);
 		UpdateViews( GetActiveView() );
 	}
 }
diff --git svx/source/editeng/impedit4.cxx svx/source/editeng/impedit4.cxx
index 684177c..8d75507 100644
--- svx/source/editeng/impedit4.cxx
+++ svx/source/editeng/impedit4.cxx
@@ -1155,7 +1155,7 @@ EditTextObject*	ImpEditEngine::CreateBinTextObject( EditSelection aSel, SfxItemP
 	// Schwelle rauf setzen, wenn Olli die Absaetze nicht mehr zerhackt!
 	if ( bAllowBigObjects && bOnlyFullParagraphs && IsFormatted() && GetUpdateMode() && ( nTextPortions >= nBigObjectStart ) )
 	{
-		XParaPortionList* pXList = new XParaPortionList( GetRefDevice(), aPaperSize.Width() );
+		XParaPortionList* pXList = new XParaPortionList( GetRefDevice(), aPaperSize.Width(), nStretchX, nStretchY );
 		pTxtObj->SetPortionInfo( pXList );
 		for ( nNode = nStartNode; nNode <= nEndNode; nNode++  )
 		{
@@ -1246,7 +1246,9 @@ EditSelection ImpEditEngine::InsertBinTextObject( BinTextObject& rTextObject, Ed
 	XParaPortionList* pPortionInfo = rTextObject.GetPortionInfo();
 
 	if ( pPortionInfo && ( (long)pPortionInfo->GetPaperWidth() == aPaperSize.Width() )
-			&& ( pPortionInfo->GetRefMapMode() == GetRefDevice()->GetMapMode() ) )
+         && ( pPortionInfo->GetRefMapMode() == GetRefDevice()->GetMapMode() ) 
+         && ( pPortionInfo->GetStretchX() == nStretchX )
+         && ( pPortionInfo->GetStretchY() == nStretchY ) )
 	{
 		if ( ( pPortionInfo->GetRefDevPtr() == (sal_uIntPtr)GetRefDevice() ) ||
 			 ( ( pPortionInfo->GetRefDevType() == OUTDEV_VIRDEV ) &&
diff --git svx/source/outliner/outliner.cxx svx/source/outliner/outliner.cxx
index 2e70966..ba9b16a 100644
--- svx/source/outliner/outliner.cxx
+++ svx/source/outliner/outliner.cxx
@@ -900,7 +900,10 @@ Font Outliner::ImpCalcBulletFont( USHORT nPara ) const
     }
 
     // #107508# Use original scale...
-	USHORT nScale = /* pEditEngine->IsFlatMode() ? DEFAULT_SCALE : */ pFmt->GetBulletRelSize();
+    USHORT nStretchX, nStretchY;
+    const_cast<Outliner*>(this)->GetGlobalCharStretching(nStretchX, nStretchY);
+    
+	USHORT nScale = pFmt->GetBulletRelSize() * nStretchY / 100;
 	ULONG nScaledLineHeight = aStdFont.GetSize().Height();
 	nScaledLineHeight *= nScale*10;
 	nScaledLineHeight /= 1000;
@@ -943,6 +946,12 @@ void Outliner::PaintBullet( USHORT nPara, const Point& rStartPos,
         BOOL bRightToLeftPara = pEditEngine->IsRightToLeft( nPara );
 
 		Rectangle aBulletArea( ImpCalcBulletArea( nPara, TRUE, FALSE ) );
+        USHORT nStretchX, nStretchY;
+        GetGlobalCharStretching(nStretchX, nStretchY);
+        aBulletArea = Rectangle( Point(aBulletArea.Left()*nStretchX/100,
+                                       aBulletArea.Top()),
+                                 Size(aBulletArea.GetWidth()*nStretchX/100,
+                                      aBulletArea.GetHeight()) );
 
         Paragraph* pPara = pParaList->GetParagraph( nPara );
         const SvxNumberFormat* pFmt = GetNumberFormat( nPara );
diff --git svx/source/sdr/attribute/sdrtextattribute.cxx svx/source/sdr/attribute/sdrtextattribute.cxx
index 527059b..e605a53 100644
--- svx/source/sdr/attribute/sdrtextattribute.cxx
+++ svx/source/sdr/attribute/sdrtextattribute.cxx
@@ -50,6 +50,7 @@ namespace drawinglayer
 			sal_Int32 aTextLowerDistance, 
 			bool bContour, 
 			bool bFitToSize, 
+			bool bAutoFit, 
 			bool bHideContour, 
 			bool bBlink, 
 			bool bScroll)
@@ -62,6 +63,7 @@ namespace drawinglayer
 			maTextLowerDistance(aTextLowerDistance),
 			mbContour(bContour),
 			mbFitToSize(bFitToSize),
+			mbAutoFit(bAutoFit),
 			mbHideContour(bHideContour),
 			mbBlink(bBlink),
 			mbScroll(bScroll)
@@ -80,6 +82,7 @@ namespace drawinglayer
 				&& getTextLowerDistance() == rCandidate.getTextLowerDistance()
 				&& isContour() == rCandidate.isContour()
 				&& isFitToSize() == rCandidate.isFitToSize()
+				&& isAutoFit() == rCandidate.isAutoFit()
 				&& isHideContour() == rCandidate.isHideContour()
 				&& isBlink() == rCandidate.isBlink()
 				&& isScroll() == rCandidate.isScroll());
diff --git svx/source/sdr/primitive2d/sdrattributecreator.cxx svx/source/sdr/primitive2d/sdrattributecreator.cxx
index b76d204..42f6900 100644
--- svx/source/sdr/primitive2d/sdrattributecreator.cxx
+++ svx/source/sdr/primitive2d/sdrattributecreator.cxx
@@ -493,7 +493,6 @@ namespace drawinglayer
 				
 				if(!bInEditMode)
 				{
-					const SdrFitToSizeType eFit = rTextObj.GetFitToSize();
 					const SdrTextAniKind eAniKind(rTextObj.GetTextAniKind());
 
 					pRetval = new attribute::SdrTextAttribute(
@@ -504,7 +503,8 @@ namespace drawinglayer
 						rTextObj.GetTextRightDistance(),
 						rTextObj.GetTextLowerDistance(),
 						((const SdrTextContourFrameItem&)rSet.Get(SDRATTR_TEXT_CONTOURFRAME)).GetValue(),
-						(SDRTEXTFIT_PROPORTIONAL == eFit || SDRTEXTFIT_ALLLINES == eFit),
+                        rTextObj.IsFitToSize(),
+                        rTextObj.IsAutoFit(),
 						((const XFormTextHideFormItem&)rSet.Get(XATTR_FORMTXTHIDEFORM)).GetValue(),
 						SDRTEXTANI_BLINK == eAniKind, 
 						SDRTEXTANI_SCROLL == eAniKind || SDRTEXTANI_ALTERNATE == eAniKind || SDRTEXTANI_SLIDE == eAniKind);
@@ -525,7 +525,7 @@ namespace drawinglayer
 				const XGradient& rGradient = ((XFillFloatTransparenceItem*)pGradientItem)->GetGradientValue();
 				const sal_uInt8 nStartLuminance(rGradient.GetStartColor().GetLuminance());
 				const sal_uInt8 nEndLuminance(rGradient.GetEndColor().GetLuminance());
-				const bool bCompletelyTransparent(0xff == nStartLuminance == nEndLuminance);
+				const bool bCompletelyTransparent(0xff == nStartLuminance && 0xff == nEndLuminance);
 
 				if(!bCompletelyTransparent)
 				{
diff --git svx/source/sdr/primitive2d/sdrdecompositiontools.cxx svx/source/sdr/primitive2d/sdrdecompositiontools.cxx
index 62461df..6b05729 100644
--- svx/source/sdr/primitive2d/sdrdecompositiontools.cxx
+++ svx/source/sdr/primitive2d/sdrdecompositiontools.cxx
@@ -267,6 +267,11 @@ namespace drawinglayer
 					// streched text in range
 					pNew = new SdrStretchTextPrimitive2D(rText.getSdrText(), aAnchorTransform);
 				}
+				else if(rText.isAutoFit())
+				{
+					// isotrophically scaled text in range
+					pNew = new SdrAutoFitTextPrimitive2D(rText.getSdrText(), aAnchorTransform, bWordWrap);
+				}
 				else // text in range
 				{
 					// build new primitive
diff --git svx/source/sdr/primitive2d/sdrtextprimitive2d.cxx svx/source/sdr/primitive2d/sdrtextprimitive2d.cxx
index 6d406fd..470bfe9 100644
--- svx/source/sdr/primitive2d/sdrtextprimitive2d.cxx
+++ svx/source/sdr/primitive2d/sdrtextprimitive2d.cxx
@@ -304,6 +304,60 @@ namespace drawinglayer
 {
 	namespace primitive2d
 	{
+		Primitive2DSequence SdrAutoFitTextPrimitive2D::createLocalDecomposition(const geometry::ViewInformation2D& aViewInformation) const
+		{
+            Primitive2DSequence aRetval;
+            const bool bCurrentSpellCheck(getSdrText().GetObject().impDecomposeAutoFitTextPrimitive(aRetval, *this, aViewInformation));
+
+            if(getLastSpellCheck() != bCurrentSpellCheck)
+            {
+                // remember last spell check state; this is part of the decomposition source data definition
+                const_cast< SdrAutoFitTextPrimitive2D* >(this)->setLastSpellCheck(bCurrentSpellCheck);
+            }
+
+            return encapsulateWithTextHierarchyBlockPrimitive2D(aRetval);
+		}
+
+		SdrAutoFitTextPrimitive2D::SdrAutoFitTextPrimitive2D(
+			const SdrText& rSdrText,
+			const ::basegfx::B2DHomMatrix& rTextRangeTransform,
+            bool bWordWrap)
+		:	SdrTextPrimitive2D(rSdrText),
+			maTextRangeTransform(rTextRangeTransform),
+            mbWordWrap(bWordWrap)
+		{
+		}
+
+		bool SdrAutoFitTextPrimitive2D::operator==(const BasePrimitive2D& rPrimitive) const
+		{
+			if(SdrTextPrimitive2D::operator==(rPrimitive))
+			{
+				const SdrBlockTextPrimitive2D& rCompare = (SdrBlockTextPrimitive2D&)rPrimitive;
+
+				return (getTextRangeTransform() == rCompare.getTextRangeTransform()
+                    && getWordWrap() == rCompare.getWordWrap());
+			}
+
+			return false;
+		}
+
+		SdrTextPrimitive2D* SdrAutoFitTextPrimitive2D::createTransformedClone(const ::basegfx::B2DHomMatrix& rTransform) const
+		{
+			return new SdrAutoFitTextPrimitive2D(getSdrText(), rTransform * getTextRangeTransform(), getWordWrap());
+		}
+
+		// provide unique ID
+		ImplPrimitrive2DIDBlock(SdrAutoFitTextPrimitive2D, PRIMITIVE2D_ID_SDRAUTOFITTEXTPRIMITIVE2D)
+
+	} // end of namespace primitive2d
+} // end of namespace drawinglayer
+
+//////////////////////////////////////////////////////////////////////////////
+
+namespace drawinglayer
+{
+	namespace primitive2d
+	{
 		Primitive2DSequence SdrStretchTextPrimitive2D::createLocalDecomposition(const geometry::ViewInformation2D& aViewInformation) const
 		{
             Primitive2DSequence aRetval;
diff --git svx/source/svdraw/svdedxv.cxx svx/source/svdraw/svdedxv.cxx
index 08e9524..b0219cb 100644
--- svx/source/svdraw/svdedxv.cxx
+++ svx/source/svdraw/svdedxv.cxx
@@ -315,7 +315,7 @@ void SdrObjEditView::ImpPaintOutlinerView(OutlinerView& rOutlView, const Rectang
 	{
 		const SdrTextObj* pText = PTR_CAST(SdrTextObj,GetTextEditObject());
 		bool bTextFrame(pText && pText->IsTextFrame());
-		bool bFitToSize(0 != (pTextEditOutliner->GetControlWord() & EE_CNTRL_STRETCHING));
+		bool bFitToSize(pText && pText->IsFitToSize());
 		bool bModifyMerk(pTextEditOutliner->IsModified()); // #43095#
 		Rectangle aBlankRect(rOutlView.GetOutputArea());
 		aBlankRect.Union(aMinTextEditArea);
@@ -384,7 +384,7 @@ void SdrObjEditView::ImpInvalidateOutlinerView(OutlinerView& rOutlView) const
 	{
 		const SdrTextObj* pText = PTR_CAST(SdrTextObj,GetTextEditObject());
 		bool bTextFrame(pText && pText->IsTextFrame());
-		bool bFitToSize(0 != (pTextEditOutliner->GetControlWord() & EE_CNTRL_STRETCHING));
+		bool bFitToSize(pText && pText->IsFitToSize());
 
 		if(bTextFrame && !bFitToSize) 
 		{
@@ -692,8 +692,7 @@ sal_Bool SdrObjEditView::SdrBeginTextEdit(
     		if ( !pTextObj->IsContourTextFrame() )
 			{
 				// FitToSize erstmal nicht mit ContourFrame
-        		SdrFitToSizeType eFit = pTextObj->GetFitToSize();
-        		if (eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES)
+        		if (pTextObj->IsFitToSize())
         			aTextRect = aAnchorRect;
 			}
 
@@ -761,8 +760,7 @@ sal_Bool SdrObjEditView::SdrBeginTextEdit(
 				// #71519#
 				if(!bExtraInvalidate)
 				{
-        			SdrFitToSizeType eFit = pTextObj->GetFitToSize();
-					if(eFit == SDRTEXTFIT_PROPORTIONAL || eFit == SDRTEXTFIT_ALLLINES)
+					if(pTextObj->IsFitToSize())
 						bExtraInvalidate = sal_True;
 				}
 
diff --git svx/source/svdraw/svdfppt.cxx svx/source/svdraw/svdfppt.cxx
index ac53ab7..4aaba2e 100644
--- svx/source/svdraw/svdfppt.cxx
+++ svx/source/svdraw/svdfppt.cxx
@@ -1113,6 +1113,7 @@ SdrObject* SdrEscherImport::ProcessObj( SvStream& rSt, DffObjData& rObjData, voi
 				}
 				aTextObj.SetDestinationInstance( (sal_uInt16)nDestinationInstance );
 
+                bool bAutoFit = false; // auto-scale text into shape box
 				switch ( aTextObj.GetInstance() )
 				{
 					case TSS_TYPE_PAGETITLE :
@@ -1120,7 +1121,7 @@ SdrObject* SdrEscherImport::ProcessObj( SvStream& rSt, DffObjData& rObjData, voi
 					case TSS_TYPE_SUBTITLE : eTextKind = OBJ_TEXT; break;
 					case TSS_TYPE_BODY :
 					case TSS_TYPE_HALFBODY :
-					case TSS_TYPE_QUARTERBODY : eTextKind = OBJ_OUTLINETEXT; break;
+					case TSS_TYPE_QUARTERBODY : eTextKind = OBJ_OUTLINETEXT; bAutoFit = true; break;
 				}
 				if ( aTextObj.GetDestinationInstance() != TSS_TYPE_TEXT_IN_SHAPE )
 				{
@@ -1175,6 +1176,15 @@ SdrObject* SdrEscherImport::ProcessObj( SvStream& rSt, DffObjData& rObjData, voi
 				}
 				pTObj->SetMergedItem( SvxFrameDirectionItem( bVerticalText ? FRMDIR_VERT_TOP_RIGHT : FRMDIR_HORI_LEFT_TOP, EE_PARA_WRITINGDIR ) );
 
+                if (bAutoFit)
+                {
+                    // disable both, defeats purpose of autofit
+                    // otherwise
+                    bAutoGrowHeight = sal_False;
+                    bAutoGrowWidth = sal_False;
+                    pTObj->SetMergedItem( SdrTextFitToSizeTypeItem(SDRTEXTFIT_AUTOFIT) );
+                }
+
 			if ( !pTObj->ISA( SdrObjCustomShape ) )
 			{
 		 		pTObj->SetMergedItem( SdrTextAutoGrowWidthItem( bAutoGrowWidth ) );
diff --git svx/source/svdraw/svdotext.cxx svx/source/svdraw/svdotext.cxx
index f52ab08..be7075b 100644
--- svx/source/svdraw/svdotext.cxx
+++ svx/source/svdraw/svdotext.cxx
@@ -145,6 +145,7 @@ SdrTextObj::SdrTextObj()
 
 	// #i25616#
 	mbSupportTextIndentingOnLineWidthChange = sal_True;
+    mbInDownScale = sal_False;
 }
 
 SdrTextObj::SdrTextObj(const Rectangle& rNewRect)
@@ -170,6 +171,7 @@ SdrTextObj::SdrTextObj(const Rectangle& rNewRect)
 
 	// #111096#
 	mbTextAnimationAllowed = sal_True;
+    mbInDownScale = sal_False;
 
 	// #108784#
 	maTextEditOffset = Point(0, 0);
@@ -200,6 +202,7 @@ SdrTextObj::SdrTextObj(SdrObjKind eNewTextKind)
 
 	// #111096#
 	mbTextAnimationAllowed = sal_True;
+    mbInDownScale = sal_False;
 
 	// #108784#
 	maTextEditOffset = Point(0, 0);
@@ -232,6 +235,7 @@ SdrTextObj::SdrTextObj(SdrObjKind eNewTextKind, const Rectangle& rNewRect)
 
 	// #111096#
 	mbTextAnimationAllowed = sal_True;
+    mbInDownScale = sal_False;
 
 	// #108784#
 	maTextEditOffset = Point(0, 0);
@@ -266,6 +270,7 @@ SdrTextObj::SdrTextObj(SdrObjKind eNewTextKind, const Rectangle& rNewRect, SvStr
 
 	// #111096#
 	mbTextAnimationAllowed = sal_True;
+    mbInDownScale = sal_False;
 
 	// #108784#
 	maTextEditOffset = Point(0, 0);
@@ -842,8 +847,7 @@ void SdrTextObj::TakeTextRect( SdrOutliner& rOutliner, Rectangle& rTextRect, FAS
 	SdrTextAniKind      eAniKind=GetTextAniKind();
 	SdrTextAniDirection eAniDirection=GetTextAniDirection();
 
-	SdrFitToSizeType eFit=GetFitToSize();
-	FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+    FASTBOOL bFitToSize(IsFitToSize());
 	FASTBOOL bContourFrame=IsContourTextFrame();
 
 	FASTBOOL bFrame=IsTextFrame();
@@ -1001,7 +1005,7 @@ OutlinerParaObject* SdrTextObj::GetEditOutlinerParaObject() const
 	return pPara;
 }
 
-void SdrTextObj::ImpSetCharStretching(SdrOutliner& rOutliner, const Rectangle& rTextRect, const Rectangle& rAnchorRect, Fraction& rFitXKorreg) const
+void SdrTextObj::ImpSetCharStretching(SdrOutliner& rOutliner, const Size& rTextSize, const Size& rShapeSize, Fraction& rFitXKorreg) const
 {
 	OutputDevice* pOut = rOutliner.GetRefDevice();
 	BOOL bNoStretching(FALSE);
@@ -1046,12 +1050,12 @@ void SdrTextObj::ImpSetCharStretching(SdrOutliner& rOutliner, const Rectangle& r
 	unsigned nLoopCount=0;
 	FASTBOOL bNoMoreLoop=FALSE;
 	long nXDiff0=0x7FFFFFFF;
-	long nWantWdt=rAnchorRect.Right()-rAnchorRect.Left();
-	long nIsWdt=rTextRect.Right()-rTextRect.Left();
+	long nWantWdt=rShapeSize.Width();
+	long nIsWdt=rTextSize.Width();
 	if (nIsWdt==0) nIsWdt=1;
 
-	long nWantHgt=rAnchorRect.Bottom()-rAnchorRect.Top();
-	long nIsHgt=rTextRect.Bottom()-rTextRect.Top();
+	long nWantHgt=rShapeSize.Height();
+	long nIsHgt=rTextSize.Height();
 	if (nIsHgt==0) nIsHgt=1;
 
 	long nXTolPl=nWantWdt/100; // Toleranz +1%
@@ -1140,8 +1144,7 @@ SdrObject* SdrTextObj::CheckHit(const Point& rPnt, USHORT nTol, const SetOfByte*
 
 	INT32 nMyTol=nTol;
 	FASTBOOL bFontwork=IsFontwork();
-	SdrFitToSizeType eFit=GetFitToSize();
-	FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+    FASTBOOL bFitToSize(IsFitToSize());
 	Rectangle aR(aRect);
 	Rectangle aAnchor2(aR);
 	Rectangle aTextRect(aR);
@@ -1374,8 +1377,7 @@ basegfx::B2DPolyPolygon SdrTextObj::TakeContour() const
 		Rectangle aR;
 		TakeTextRect(rOutliner,aR,FALSE,&aAnchor2);
 		rOutliner.Clear();
-		SdrFitToSizeType eFit=GetFitToSize();
-		FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+        FASTBOOL bFitToSize(IsFitToSize());
 		if (bFitToSize) aR=aAnchor2;
 		Polygon aPol(aR);
 		if (aGeo.nDrehWink!=0) RotatePoly(aPol,aR.TopLeft(),aGeo.nSin,aGeo.nCos);
@@ -1492,8 +1494,7 @@ void SdrTextObj::ImpSetupDrawOutlinerForPaint( FASTBOOL 		bContourFrame,
     if (!bContourFrame)
     {
         // FitToSize erstmal nicht mit ContourFrame
-        SdrFitToSizeType eFit=GetFitToSize();
-        if (eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES)
+        if (IsFitToSize() || IsAutoFit())
         {
             ULONG nStat=rOutliner.GetControlWord();
             nStat|=EE_CNTRL_STRETCHING|EE_CNTRL_AUTOPAGESIZE;
@@ -1507,15 +1508,75 @@ void SdrTextObj::ImpSetupDrawOutlinerForPaint( FASTBOOL 		bContourFrame,
     if (!bContourFrame)
     {
         // FitToSize erstmal nicht mit ContourFrame
-        SdrFitToSizeType eFit=GetFitToSize();
-        if (eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES)
+        if (IsFitToSize())
         {
-            ImpSetCharStretching(rOutliner,rTextRect,rAnchorRect,rFitXKorreg);
+            ImpSetCharStretching(rOutliner,rTextRect.GetSize(),rAnchorRect.GetSize(),rFitXKorreg);
             rPaintRect=rAnchorRect;
         }
+        else if (IsAutoFit())
+        {
+            ImpAutoFitText(rOutliner);
+        }
     }
 }
 
+void SdrTextObj::ImpAutoFitText( SdrOutliner& rOutliner ) const
+{
+    const Size aShapeSize=GetSnapRect().GetSize();
+    ImpAutoFitText( rOutliner, 
+                    Size(aShapeSize.Width()-GetTextLeftDistance()-GetTextRightDistance(),
+                         aShapeSize.Height()-GetTextUpperDistance()-GetTextLowerDistance()), 
+                    IsVerticalWriting() );
+}
+
+void SdrTextObj::ImpAutoFitText( SdrOutliner& rOutliner, const Size& rTextSize, bool bIsVerticalWriting )
+{
+    // EditEngine formatting is unstable enough for
+    // line-breaking text that we need some more samples
+
+    // loop early-exits if we detect an already attained value
+    USHORT nMinStretchX=0, nMinStretchY=0;
+    USHORT aOldStretchXVals[]={0,0,0,0,0,0,0,0,0,0};
+    const size_t aStretchArySize=sizeof(aOldStretchXVals)/sizeof(*aOldStretchXVals);
+    for(int i=0; i<aStretchArySize; ++i)
+    {
+        const Size aCurrTextSize = rOutliner.CalcTextSize();
+        double fFactor(1.0);
+        if( bIsVerticalWriting )
+            fFactor = double(rTextSize.Width())/aCurrTextSize.Width();
+        else
+            fFactor = double(rTextSize.Height())/aCurrTextSize.Height();
+
+        USHORT nCurrStretchX, nCurrStretchY;
+        rOutliner.GetGlobalCharStretching(nCurrStretchX, nCurrStretchY);
+
+        if (fFactor >= 1.0 )
+        {
+            // resulting text area fits into available shape rect -
+            // err on the larger streching, to optimally fill area
+            nMinStretchX = std::max(nMinStretchX,nCurrStretchX);
+            nMinStretchY = std::max(nMinStretchY,nCurrStretchY);
+        }
+
+        aOldStretchXVals[i] = nCurrStretchX;
+        if( std::find(aOldStretchXVals, aOldStretchXVals+i, nCurrStretchX) != aOldStretchXVals+i )
+            break; // same value already attained once; algo is looping, exit
+
+        if (fFactor < 1.0 || (fFactor >= 1.0 && nCurrStretchX != 100))
+        {
+            nCurrStretchX = sal::static_int_cast<USHORT>(nCurrStretchX*fFactor);
+            nCurrStretchY = sal::static_int_cast<USHORT>(nCurrStretchY*fFactor);
+            rOutliner.SetGlobalCharStretching(std::min(USHORT(100),nCurrStretchX),
+                                              std::min(USHORT(100),nCurrStretchY));
+            OSL_TRACE("SdrTextObj::onEditOutlinerStatusEvent(): zoom is %d", nCurrStretchX);
+        }
+    }
+
+    OSL_TRACE("---- SdrTextObj::onEditOutlinerStatusEvent(): final zoom is %d ----", nMinStretchX);
+    rOutliner.SetGlobalCharStretching(std::min(USHORT(100),nMinStretchX),
+                                      std::min(USHORT(100),nMinStretchY));
+}
+
 void SdrTextObj::SetupOutlinerFormatting( SdrOutliner& rOutl, Rectangle& rPaintRect ) const
 {
     ImpInitDrawOutliner( rOutl );
@@ -2107,6 +2168,17 @@ bool SdrTextObj::IsTextAnimationAllowed() const
 	return mbTextAnimationAllowed;
 }
 
+FASTBOOL SdrTextObj::IsAutoFit() const
+{
+    return GetFitToSize()==SDRTEXTFIT_AUTOFIT;
+}
+
+FASTBOOL SdrTextObj::IsFitToSize() const
+{
+    const SdrFitToSizeType eFit=GetFitToSize();
+    return (eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+}
+
 void SdrTextObj::SetTextAnimationAllowed(sal_Bool bNew)
 {
 	if(mbTextAnimationAllowed != bNew)
@@ -2124,13 +2196,21 @@ void SdrTextObj::onEditOutlinerStatusEvent( EditStatus* pEditStatus )
 	const bool bGrowY=(nStat & EE_STAT_TEXTHEIGHTCHANGED) !=0;
     if(bTextFrame && (bGrowX || bGrowY))
 	{
-		const bool bAutoGrowHgt= bTextFrame && IsAutoGrowHeight();
-		const bool bAutoGrowWdt= bTextFrame && IsAutoGrowWidth();
-
-	    if ((bGrowX && bAutoGrowWdt) || (bGrowY && bAutoGrowHgt))
+	    if ((bGrowX && IsAutoGrowWidth()) || (bGrowY && IsAutoGrowHeight()))
 		{
 			AdjustTextFrameWidthAndHeight();
 		}
+        else if (IsAutoFit() && !mbInDownScale)
+        {
+            OSL_ASSERT(pEdtOutl);
+            mbInDownScale = sal_True;
+
+            // sucks that we cannot disable paints via
+            // pEdtOutl->SetUpdateMode(FALSE) - but EditEngine skips
+            // formatting as well, then.
+            ImpAutoFitText(*pEdtOutl);
+            mbInDownScale = sal_False;
+        }
 	}
 }
 
diff --git svx/source/svdraw/svdotextdecomposition.cxx svx/source/svdraw/svdotextdecomposition.cxx
index 554070d..fac3231 100644
--- svx/source/svdraw/svdotextdecomposition.cxx
+++ svx/source/svdraw/svdotextdecomposition.cxx
@@ -871,11 +871,11 @@ bool SdrTextObj::impDecomposeStretchTextPrimitive(
 	basegfx::B2DHomMatrix aNewTransformA;
 	basegfx::B2DHomMatrix aNewTransformB;
 
-	// calculate global char stretching scale parameters. Use non-mirrored sizes
-	// to layout without mirroring
-	const double fScaleX(fabs(aScale.getX()) / aOutlinerScale.getX());
-	const double fScaleY(fabs(aScale.getY()) / aOutlinerScale.getY());
-	rOutliner.SetGlobalCharStretching((sal_Int16)FRound(fScaleX * 100.0), (sal_Int16)FRound(fScaleY * 100.0));
+    // calculate global char stretching scale parameters. Use non-mirrored sizes
+    // to layout without mirroring. Scale anisotrophically
+    const double fScaleX(fabs(aScale.getX()) / aOutlinerScale.getX());
+    const double fScaleY(fabs(aScale.getY()) / aOutlinerScale.getY());
+    rOutliner.SetGlobalCharStretching((sal_Int16)FRound(fScaleX * 100.0), (sal_Int16)FRound(fScaleY * 100.0));
 
 	// mirroring. We are now in aAnchorTextRange sizes. When mirroring in X and Y,
 	// move the null point which was top left to bottom right.
@@ -902,6 +902,136 @@ bool SdrTextObj::impDecomposeStretchTextPrimitive(
     return false;
 }
 
+bool SdrTextObj::impDecomposeAutoFitTextPrimitive(
+	drawinglayer::primitive2d::Primitive2DSequence& rTarget, 
+	const drawinglayer::primitive2d::SdrAutoFitTextPrimitive2D& rSdrAutofitTextPrimitive,
+	const drawinglayer::geometry::ViewInformation2D& aViewInformation) const
+{
+    // decompose matrix to have position and size of text
+	basegfx::B2DVector aScale, aTranslate;
+	double fRotate, fShearX;
+	rSdrAutofitTextPrimitive.getTextRangeTransform().decompose(aScale, aTranslate, fRotate, fShearX);
+
+	// use B2DRange aAnchorTextRange for calculations
+	basegfx::B2DRange aAnchorTextRange(aTranslate);
+	aAnchorTextRange.expand(aTranslate + aScale);
+
+	// prepare outliner
+	const SfxItemSet& rTextItemSet = rSdrAutofitTextPrimitive.getSdrText().GetItemSet();
+	SdrOutliner& rOutliner = ImpGetDrawOutliner();
+	SdrTextVertAdjust eVAdj = GetTextVerticalAdjust(rTextItemSet);
+	SdrTextHorzAdjust eHAdj = GetTextHorizontalAdjust(rTextItemSet);
+	const sal_uInt32 nOriginalControlWord(rOutliner.GetControlWord());
+	const Size aNullSize;
+
+	// set visualizing page at Outliner; needed e.g. for PageNumberField decomposition
+	rOutliner.setVisualizedPage(GetSdrPageFromXDrawPage(aViewInformation.getVisualizedPage()));
+
+	rOutliner.SetControlWord(nOriginalControlWord|EE_CNTRL_AUTOPAGESIZE|EE_CNTRL_STRETCHING);
+	rOutliner.SetMinAutoPaperSize(aNullSize);
+	rOutliner.SetMaxAutoPaperSize(Size(1000000,1000000));
+
+	// add one to rage sizes to get back to the old Rectangle and outliner measurements
+	const sal_uInt32 nAnchorTextWidth(FRound(aAnchorTextRange.getWidth() + 1L));
+	const sal_uInt32 nAnchorTextHeight(FRound(aAnchorTextRange.getHeight() + 1L));
+	const OutlinerParaObject* pOutlinerParaObject = rSdrAutofitTextPrimitive.getSdrText().GetOutlinerParaObject();
+	OSL_ENSURE(pOutlinerParaObject, "impDecomposeBlockTextPrimitive used with no OutlinerParaObject (!)");
+	const bool bVerticalWritintg(pOutlinerParaObject->IsVertical());
+	const Size aAnchorTextSize(Size(nAnchorTextWidth, nAnchorTextHeight));
+
+    if((rSdrAutofitTextPrimitive.getWordWrap() || IsTextFrame()))
+    {
+        rOutliner.SetMaxAutoPaperSize(aAnchorTextSize);
+    }
+
+    if(SDRTEXTHORZADJUST_BLOCK == eHAdj && !bVerticalWritintg)
+    {
+        rOutliner.SetMinAutoPaperSize(Size(nAnchorTextWidth, 0));
+    }
+
+    if(SDRTEXTVERTADJUST_BLOCK == eVAdj && bVerticalWritintg)
+    {
+        rOutliner.SetMinAutoPaperSize(Size(0, nAnchorTextHeight));
+    }
+
+    rOutliner.SetPaperSize(aNullSize);
+    rOutliner.SetUpdateMode(true);
+    rOutliner.SetText(*pOutlinerParaObject);
+    ImpAutoFitText(rOutliner,aAnchorTextSize,bVerticalWritintg);
+
+	// now get back the layouted text size from outliner
+	const Size aOutlinerTextSiz(rOutliner.GetPaperSize());
+	const basegfx::B2DVector aOutlinerScale(aOutlinerTextSiz.Width(), aOutlinerTextSiz.Height());
+	basegfx::B2DVector aAdjustTranslate(0.0, 0.0);
+
+	// correct horizontal translation using the now known text size
+	if(SDRTEXTHORZADJUST_CENTER == eHAdj || SDRTEXTHORZADJUST_RIGHT == eHAdj)
+	{
+		const double fFree(aAnchorTextRange.getWidth() - aOutlinerScale.getX());
+
+		if(SDRTEXTHORZADJUST_CENTER == eHAdj)
+		{
+			aAdjustTranslate.setX(fFree / 2.0);
+		}
+
+		if(SDRTEXTHORZADJUST_RIGHT == eHAdj)
+		{
+			aAdjustTranslate.setX(fFree);
+		}
+	}
+
+	// correct vertical translation using the now known text size
+	if(SDRTEXTVERTADJUST_CENTER == eVAdj || SDRTEXTVERTADJUST_BOTTOM == eVAdj)
+	{
+		const double fFree(aAnchorTextRange.getHeight() - aOutlinerScale.getY());
+
+		if(SDRTEXTVERTADJUST_CENTER == eVAdj)
+		{
+			aAdjustTranslate.setY(fFree / 2.0);
+		}
+		
+		if(SDRTEXTVERTADJUST_BOTTOM == eVAdj)
+		{
+			aAdjustTranslate.setY(fFree);
+		}
+	}
+
+	// prepare matrices to apply to newly created primitives. aNewTransformA
+	// will get coordinates in aOutlinerScale size and positive in X, Y.
+	basegfx::B2DHomMatrix aNewTransformA;
+	basegfx::B2DHomMatrix aNewTransformB;
+
+	// translate relative to given primitive to get same rotation and shear
+	// as the master shape we are working on. For vertical, use the top-right
+	// corner
+	const double fStartInX(bVerticalWritintg ? aAdjustTranslate.getX() + aOutlinerScale.getX() : aAdjustTranslate.getX());
+	aNewTransformA.translate(fStartInX, aAdjustTranslate.getY());
+
+	// mirroring. We are now in aAnchorTextRange sizes. When mirroring in X and Y,
+	// move the null point which was top left to bottom right.
+	const bool bMirrorX(basegfx::fTools::less(aScale.getX(), 0.0));
+	const bool bMirrorY(basegfx::fTools::less(aScale.getY(), 0.0));
+	aNewTransformB.scale(bMirrorX ? -1.0 : 1.0, bMirrorY ? -1.0 : 1.0);
+
+	// in-between the translations of the single primitives will take place. Afterwards,
+	// the object's transformations need to be applied
+	aNewTransformB.shearX(fShearX);
+	aNewTransformB.rotate(fRotate);
+	aNewTransformB.translate(aTranslate.getX(), aTranslate.getY());
+
+	// now break up text primitives.
+	impTextBreakupHandler aConverter(rOutliner);
+	aConverter.decomposeBlockTextPrimitive(aNewTransformA, aNewTransformB);
+
+	// cleanup outliner
+	rOutliner.Clear();
+	rOutliner.setVisualizedPage(0);
+    rOutliner.SetControlWord(nOriginalControlWord);
+
+	rTarget = aConverter.getPrimitive2DSequence();
+    return false;
+}
+
 //////////////////////////////////////////////////////////////////////////////
 // timing generators
 #define ENDLESS_LOOP	(0xffffffff)
diff --git svx/source/svdraw/svdotxat.cxx svx/source/svdraw/svdotxat.cxx
index 67b6cc2..f3e4564 100644
--- svx/source/svdraw/svdotxat.cxx
+++ svx/source/svdraw/svdotxat.cxx
@@ -82,8 +82,7 @@ FASTBOOL SdrTextObj::AdjustTextFrameWidthAndHeight(Rectangle& rR, FASTBOOL bHgt,
 {
 	if (bTextFrame && pModel!=NULL && !rR.IsEmpty())
 	{
-		SdrFitToSizeType eFit=GetFitToSize();
-		FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+        FASTBOOL bFitToSize(IsFitToSize());
 		FASTBOOL bWdtGrow=bWdt && IsAutoGrowWidth();
 		FASTBOOL bHgtGrow=bHgt && IsAutoGrowHeight();
 		SdrTextAniKind eAniKind=GetTextAniKind();
diff --git svx/source/svdraw/svdotxed.cxx svx/source/svdraw/svdotxed.cxx
index 94c8a20..741b793 100644
--- svx/source/svdraw/svdotxed.cxx
+++ svx/source/svdraw/svdotxed.cxx
@@ -75,15 +75,17 @@ sal_Bool SdrTextObj::BegTextEdit(SdrOutliner& rOutl)
 	rOutl.Init( nOutlinerMode );
 	rOutl.SetRefDevice( pModel->GetRefDevice() );
 
-	SdrFitToSizeType eFit=GetFitToSize();
-	FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+    FASTBOOL bFitToSize(IsFitToSize());
 	FASTBOOL bContourFrame=IsContourTextFrame();
 	ImpSetTextEditParams();
 
 	if (!bContourFrame) {
 		ULONG nStat=rOutl.GetControlWord();
 		nStat|=EE_CNTRL_AUTOPAGESIZE;
-		if (bFitToSize) nStat|=EE_CNTRL_STRETCHING; else nStat&=~EE_CNTRL_STRETCHING;
+		if (bFitToSize || IsAutoFit()) 
+            nStat|=EE_CNTRL_STRETCHING; 
+        else 
+            nStat&=~EE_CNTRL_STRETCHING;
 		rOutl.SetControlWord(nStat);
 	}
 
@@ -119,8 +121,12 @@ sal_Bool SdrTextObj::BegTextEdit(SdrOutliner& rOutl)
 		TakeTextRect(rOutl, aTextRect, FALSE, 
 			&aAnchorRect/* #97097# give TRUE here, not FALSE */);
 		Fraction aFitXKorreg(1,1);
-		ImpSetCharStretching(rOutl,aTextRect,aAnchorRect,aFitXKorreg);
+		ImpSetCharStretching(rOutl,aTextRect.GetSize(),aAnchorRect.GetSize(),aFitXKorreg);
 	}
+    else if (IsAutoFit())
+    {
+        ImpAutoFitText(rOutl);
+    }
 
 	if(pOutlinerParaObject)
 	{
@@ -146,8 +152,7 @@ sal_Bool SdrTextObj::BegTextEdit(SdrOutliner& rOutl)
 
 void SdrTextObj::TakeTextEditArea(Size* pPaperMin, Size* pPaperMax, Rectangle* pViewInit, Rectangle* pViewMin) const
 {
-	SdrFitToSizeType eFit=GetFitToSize();
-	FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+    FASTBOOL bFitToSize(IsFitToSize());
 	Size aPaperMin,aPaperMax;
 	Rectangle aViewInit;
 	TakeTextAnchorRect(aViewInit);
diff --git svx/source/svdraw/svdotxtr.cxx svx/source/svdraw/svdotxtr.cxx
index a1a82f6..08f73bf 100644
--- svx/source/svdraw/svdotxtr.cxx
+++ svx/source/svdraw/svdotxtr.cxx
@@ -81,9 +81,6 @@ void SdrTextObj::NbcSetSnapRect(const Rectangle& rRect)
 		if (bTextFrame && (pModel==NULL || !pModel->IsPasteResize())) { // #51139#
 			if (nTWdt0!=nTWdt1 && IsAutoGrowWidth() ) NbcSetMinTextFrameWidth(nTWdt1);
 			if (nTHgt0!=nTHgt1 && IsAutoGrowHeight()) NbcSetMinTextFrameHeight(nTHgt1);
-			if (GetFitToSize()==SDRTEXTFIT_RESIZEATTR) {
-				NbcResizeTextAttributes(Fraction(nTWdt1,nTWdt0),Fraction(nTHgt1,nTHgt0));
-			}
 			NbcAdjustTextFrameWidthAndHeight();
 		}
 		ImpCheckShear();
@@ -109,9 +106,6 @@ void SdrTextObj::NbcSetLogicRect(const Rectangle& rRect)
 	if (bTextFrame) {
 		if (nTWdt0!=nTWdt1 && IsAutoGrowWidth() ) NbcSetMinTextFrameWidth(nTWdt1);
 		if (nTHgt0!=nTHgt1 && IsAutoGrowHeight()) NbcSetMinTextFrameHeight(nTHgt1);
-		if (GetFitToSize()==SDRTEXTFIT_RESIZEATTR) {
-			NbcResizeTextAttributes(Fraction(nTWdt1,nTWdt0),Fraction(nTHgt1,nTHgt0));
-		}
 		NbcAdjustTextFrameWidthAndHeight();
 	}
 	SetRectsDirty();
@@ -225,9 +219,6 @@ void SdrTextObj::NbcResize(const Point& rRef, const Fraction& xFact, const Fract
 	if (bTextFrame && (pModel==NULL || !pModel->IsPasteResize())) { // #51139#
 		if (nTWdt0!=nTWdt1 && IsAutoGrowWidth() ) NbcSetMinTextFrameWidth(nTWdt1);
 		if (nTHgt0!=nTHgt1 && IsAutoGrowHeight()) NbcSetMinTextFrameHeight(nTHgt1);
-		if (GetFitToSize()==SDRTEXTFIT_RESIZEATTR) {
-			NbcResizeTextAttributes(Fraction(nTWdt1,nTWdt0),Fraction(nTHgt1,nTHgt0));
-		}
 		NbcAdjustTextFrameWidthAndHeight();
 	}
 	ImpCheckShear();
diff --git svx/source/svdraw/svdtxhdl.cxx svx/source/svdraw/svdtxhdl.cxx
index d1a4401..7450c5a 100644
--- svx/source/svdraw/svdtxhdl.cxx
+++ svx/source/svdraw/svdtxhdl.cxx
@@ -84,8 +84,7 @@ void ImpTextPortionHandler::ConvertToPathObj(SdrObjGroup& rGroup, FASTBOOL bPoly
 
 	Rectangle aAnchorRect;
 	Rectangle aTextRect;
-	SdrFitToSizeType eFit=rTextObj.GetFitToSize();
-	FASTBOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+    FASTBOOL bFitToSize(rTextObj.IsFitToSize());
 	// Bei TakeTextRect wird u.a. auch der Text in
 	// den Outliner gesteckt
 	rTextObj.TakeTextRect(rOutliner,aTextRect,FALSE,&aAnchorRect);
diff --git svx/source/svdraw/svdview.cxx svx/source/svdraw/svdview.cxx
index 9c42be1..1c281c8 100644
--- svx/source/svdraw/svdview.cxx
+++ svx/source/svdraw/svdview.cxx
@@ -504,8 +504,7 @@ SdrHitKind SdrView::PickAnything(const Point& rLogicPos, SdrViewEvent& rVEvt) co
 				Point aTemporaryTextRelativePosition(aLocalLogicPosition - aTextRect.TopLeft());
 				
 				// FitToSize berueksichtigen
-				SdrFitToSizeType eFit=pTextObj->GetFitToSize();
-				BOOL bFitToSize=(eFit==SDRTEXTFIT_PROPORTIONAL || eFit==SDRTEXTFIT_ALLLINES);
+                BOOL bFitToSize(pTextObj->IsFitToSize());
 				if (bFitToSize) {
 					Fraction aX(aTextRect.GetWidth()-1,aAnchor.GetWidth()-1);
 					Fraction aY(aTextRect.GetHeight()-1,aAnchor.GetHeight()-1);
diff --git xmloff/source/draw/sdpropls.cxx xmloff/source/draw/sdpropls.cxx
index 0182e28..9df993c 100644
--- xmloff/source/draw/sdpropls.cxx
+++ xmloff/source/draw/sdpropls.cxx
@@ -624,8 +624,8 @@ SvXMLEnumMapEntry __READONLY_DATA pXML_FitToSize_Enum[] =
 {
 	{ XML_FALSE,		drawing::TextFitToSizeType_NONE },
 	{ XML_TRUE, 		drawing::TextFitToSizeType_PROPORTIONAL },
-	{ XML_TRUE,	    	drawing::TextFitToSizeType_ALLLINES },
-	{ XML_TRUE,		    drawing::TextFitToSizeType_RESIZEATTR },
+	{ XML_ALL,	    	drawing::TextFitToSizeType_ALLLINES },
+	{ XML_SHRINK_TO_FIT,drawing::TextFitToSizeType_AUTOFIT },
 	{ XML_TOKEN_INVALID, 0 }
 };
 
